### DISCUSSION TITLE
Restart on poly_grain_growth_2D_eldrforce.i

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#17441] [Q&A General] [answered]
Dear MOOSE experts,
Recently I am using [the restart function of moose](https://mooseframework.inl.gov/application_usage/restart_recover.html), based on the examples of [power_law_creep_restart1.i and power_law_creep_restart2.i ](https://github.com/idaholab/moose/tree/next/modules/combined/test/tests/power_law_creep)to modify the [combined/poly_grain_growth_2D_eldrforce.i](https://github.com/idaholab/moose/blob/next/modules/combined/examples/phase_field-mechanics/poly_grain_growth_2D_eldrforce.i) to get [poly_grain_growth_2D_eldrforce_restart1.i and poly_grain_restart2.i](https://github.com/PengWei97/monkey/tree/main/problem/restart). I hope the restart function can work on this example. But an error is displayed in GrainTracker, and the error displayed to the terminal is as follows, [poly_grain_growth_2D_eldrforce_restart.log](https://github.com/PengWei97/monkey/blob/main/problem/restart/poly_grain_growth_2D_eldrforce_restart.log)

```
Grain Tracker Status:
…
inished inside of GrainTracker
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 64255 RUNNING AT xia-virtual-machine
=   EXIT CODE: 139
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestion
```

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thank you
Wei Peng

**Comment by GiudGiud** [comment, marked as answer]
Hello

Could you please run this in a debugger (gdb or lldb) ? Seg faults should be fairly easy to track and will give us more information about what is failing. 
Much more information will be displayed if you do this with a debug build of moose (METHOD=dbg make to build)

Guillaume

**Reply by PengWei97** [reply]
> Hello
> 
> Could you please run this in a debugger (gdb or lldb) ? Seg faults should be fairly easy to track and will give us more information about what is failing.
> Much more information will be displayed if you do this with a debug build of moose (METHOD=dbg make to build)
> 
> Guillaume

Hi, Guillaume. Thank you for your help.
I used the dbg debugger to run my poly_grain_growth_2D_eldrforce_restart2.i file, and its errors are shown as follows,
```
Finished inside of GrainTracker

Assertion'grain id <grain data.size()' failed
Requested data for invalid grain index.
at /home/xia/projects/moose/modules/phase field/build/header _symlinks/GrainDataTracker.h,line 46
Stack frames: 22
```

This error is displayed in the GrainDataTracker.h file. To be honest, I am not sure how to solve this error. Would you have any suggestions?

**Reply by GiudGiud** [reply]
Thanks. Could you please provide the backtrace as well? For this, once the debug build has completed in the debugger, type `bt`

@permcody should be able to help you then for grain growth problems

**Reply by PengWei97** [reply]
Thank you very much for your reply. 
According to your suggestion, once the debug build has completed in the debugger, type `b MPI_Abort`. 
The information displayed is in the [debug_informatin.log](https://github.com/PengWei97/monkey/blob/main/problem/restart/debug_informatin.log) file.

**Reply by permcody** [reply]
The restart issue may not be the underlying issue here. I seem to recall this particular input needed some other fixups. Are you able to run the simulation to completion without using the restart capability? If so, it may just be a question of going through the specific objects in this file and making sure any stateful data is properly declared as "restartable".

**Reply by PengWei97** [reply]
Thanks, @permcody 
Yes, if the restart capability is not applicable, that is the Problem module is deleted from the input file, the simulation can be run. And, you said to ensure that any stateful data is properly declared as "restartable". I don't really understand how to do it. 
I think this problem should appear on the graintracker object, because it is time to use the restart capability to show that it is due to `"Assertion'grain id <grain data.size()' failed"`

**Reply by PengWei97** [reply]
hi @permcody,

I searched for the keyword ‘restartable’ on the moose website and found the following. `declareRestartableData()` is used to declare that the data is ‘Restartable’. Details as follow,
![image](https://user-images.githubusercontent.com/56183775/114373988-ba2f9b00-9bb5-11eb-8fa0-3b3d4c775c31.png)

Then I searched the variables declared as Restartable in [GrainTracker.C](https://github.com/idaholab/moose/blob/6437ec83fe011ae67c7b71b4210b426db6ca320b/modules/phase_field/src/postprocessors/GrainTracker.C) and found that there are two variables. Respectively are

> ./GrainTracker.C: _first_time(declareRestartableData<bool>("first_time", true)),
> ./GrainTracker.C: _max_curr_grain_id(declareRestartableData<unsigned int>("max_curr_grain_id", invalid_id)),

According to your suggestion, should I look for the variable that cannot be restarted in GrainTracker.C, that is, the variable needed for restart, and finally declare it as ‘Restartable’.

Best Regards,
wei

**Reply by GiudGiud** [reply]
Is this solved?
From the debug log it seems an assert is failing.
 
Are both items compared in this assert restarted (grain_id and grain_data I think)? You may want to print them after the restart

**Reply by PengWei97** [reply]
Thanks, @GiudGiud
Yes, this problem has not been resolved. So I cannot use the restart function on this coupling example.

> From the debug log it seems an assert is failing.

Please forgive me for being a rookie for using the moose debugging environment, and I can only set these 5 breakpoints in the mian function. The following is the information I think is more important when debugging. For detailed information, please read [restart02.md](https://github.com/PengWei97/monkey/blob/main/problem/restart/restart02.md).

Hope the following information will help me to solve this problem,

```bash
Assertion `grain_id < _grain_data.size()' failed
Requested data for invalid grain index.
at /home/xia/projects/moose/modules/phase_field/build/header_symlinks/GrainDataTracker.h, line 46
Stack frames: 22
0: libMesh::print_trace(std::ostream&)
1: GrainDataTracker<RankFourTensorTempl<double> >::getData(unsigned int) const
2: ComputePolycrystalElasticityTensor::computeQpElasticityTensor()
3: ComputeElasticityTensorBaseTempl<false>::computeQpProperties()
4: Material::computeProperties()
5: MaterialData::reinit(std::__debug::vector<std::shared_ptr<MaterialBase>, std::allocator<std::shared_ptr<MaterialBase> > > const&, bool)
6: FEProblemBase::reinitMaterials(unsigned short, unsigned int, bool)
7: ComputeElemAuxVarsThread<AuxKernelTempl<double> >::onElement(libMesh::Elem const*)
8: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
9: /home/xia/projects/moose/framework/libmoose-dbg.so.0(+0x3ab1c4e) [0x7ffff605dc4e]
10: void AuxiliarySystem::computeElementalVarsHelper<AuxKernelTempl<double> >(MooseObjectWarehouse<AuxKernelTempl<double> > const&, std::__debug::vector<std::__debug::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> >, std::allocator<std::__debug::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> > > > const&, unsigned int)
11: AuxiliarySystem::computeElementalVars(MooseEnumItem)
12: AuxiliarySystem::compute(MooseEnumItem)
13: FEProblemBase::computeAuxiliaryKernels(MooseEnumItem const&)
14: FEProblemBase::execute(MooseEnumItem const&)
15: FEProblemBase::initialSetup()
16: Transient::init()
17: MooseApp::executeExecutioner()
18: MooseApp::run()
19: /home/xia/projects/kunpeng/kunpeng-dbg(+0x1434b) [0x55555556834b]
20: __libc_start_main
21: /home/xia/projects/kunpeng/kunpeng-dbg(+0x13189) [0x555555567189]
[0] /home/xia/projects/moose/modules/phase_field/build/header_symlinks/GrainDataTracker.h, line 46, compiled May  7 2021 at 17:40:01
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
---------------------------------------------------------------------------------
| Time:           Fri May  7 20:02:09 2021                                        |
| OS:             Linux                                                           |
| HostName:       xia-virtual-machine                                             |
| OS Release:     5.8.0-50-generic                                                |
| OS Version:     #56~20.04.1-Ubuntu SMP Mon Apr 12 21:46:35 UTC 2021             |
| Machine:        x86_64                                                          |
| Username:       xia                                                             |
| Configuration:  ../configure  '--enable-silent-rules'                           |
|  '--enable-unique-id'                                                           |
|  '--disable-warnings'                                                           |
|  '--enable-glibcxx-debugging'                                                   |
|  '--with-thread-model=openmp'                                                   |
|  '--disable-maintainer-mode'                                                    |
|  '--enable-petsc-hypre-required'                                                |
|  '--enable-metaphysicl-required'                                                |
|  '--prefix=/home/xia/miniconda3/envs/moose/libmesh'                             |
|  '--with-vtk-lib=/home/xia/miniconda3/envs/moose/libmesh-vtk/lib'               |
|  '--with-vtk-include=/home/xia/miniconda3/envs/moose/libmesh-vtk/include/vtk-6.3'|
|  '--with-methods=opt oprof devel dbg'                                           |
|  '--without-gdb-command'                                                        |
|  'build_alias=x86_64-conda-linux-gnu'                                           |
|  'host_alias=x86_64-conda-linux-gnu'                                            |
|  'PETSC_DIR=/home/xia/miniconda3/envs/moose'                                    |
|  'CC=mpicc'                                                                     |
|  'CFLAGS=-march=nocona -mtune=haswell'                                          |
|  'LDFLAGS=-Wl,-S'                                                               |
|  'FC=mpifort'                                                                   |
|  'F77=mpifort'                                                                  |
|  'CXX=mpicxx'                                                                   |
|  'CXXFLAGS=-march=nocona -mtune=haswell'                                        |
|  'CPP=/home/xia/miniconda3/envs/moose/bin/x86_64-conda-linux-gnu-cpp'           |
 ---------------------------------------------------------------------------------
 ------------------------------------------------------------------------------------------------------------
| kunpeng Performance: Alive time=868.405, Active time=0.000206                                              |
 ------------------------------------------------------------------------------------------------------------
| Event                         nCalls     Total Time  Avg Time    Total Time  Avg Time    % of Active Time  |
|                                          w/o Sub     w/o Sub     With Sub    With Sub    w/o S    With S   |
|------------------------------------------------------------------------------------------------------------|
|                                                                                                            |
|                                                                                                            |
| PolycrystalICTools                                                                                         |
|   assignOpsToGrains()         1          0.0002      0.000206    0.0002      0.000206    100.00   100.00   |
 ------------------------------------------------------------------------------------------------------------
| Totals:                       1          0.0002                                          100.00            |
 ------------------------------------------------------------------------------------------------------------

--Type <RET> for more, q to quit, c to continue without paging--

```

> Are both items compared in this assert restarted (grain_id and grain_data I think)? You may want to print them after the restart

Please forgive me for not understanding the meaning of this sentence. Is it necessary to set these two data to "restartable"?

And according to the suggestion from @permcody , `making sure any stateful data is properly declared as "restartable"`. The terminal shows that the error occurred in /home/xia/projects/moose/modules/phase_field/build/header_symlinks/GrainDataTracker.h
```c++
template <typename T>
const T &
GrainDataTracker<T>::getData(unsigned int grain_id) const
{
   mooseAssert(grain_id <_grain_data.size(), "Requested data for invalid grain index.");
   return _grain_data[grain_id];
}
```
Do I need to declare grain_id and _grain_data as "restartable"?  Feel free to try to do this, but I don't know how to declare a variable to be "restartable".

Best Regards,
wei

**Reply by GiudGiud** [reply]
grain_id is the function parameter so it's not the issue. / it does not need to be declared restartable. 

_grain_data however is a good candidate to be restartable. It's initializaed by newGrainCreated in the initial simulation and I dont know but I suspect it's not called when restarting. 
Could you try that?

Sorry this is not my area of expertise in the code.

**Reply by PengWei97** [reply]
Thanks very much for your enthusiastic answer, @GiudGiud 
I think what you said is correct. I performed the `restart` operation on `grain_growth_2D_graintracker.i`, and it proved that it was operational. However, the problematic `poly_grain_growth_2D_eldrforce.i` used a derived class of `GrainDataTracker.h`, namely `GrainTrackerElasticity`, and `GrainDataTracker.h` is a template class that inherits `GrainDataTracker`.

Therefore, I also think that restart is interrupted somewhere in `GrainDataTracker.h`.

Finally,I will try my best to solve this problem, and hope you can provide more opinions for moose beginners like me.

Thanks,
wei

**Reply by GiudGiud** [reply]
Any luck?

**Reply by PengWei97** [reply]
No, there is only a preliminary idea here. According to the error displayed during compilation, I think it may be caused by `_grain_data` not being declared as "restartable", so I plan to try to declare `_grain_data` as "restartable" to see if this problem can be solved.

However, I don't know how to set a variable to "restartable", even after browsing [DataIO](https://mooseframework.inl.gov/source/restart/DataIO.html#declareRestartableData), can you provide a specific example so that I can imitate it to complete this operation.

**Reply by GiudGiud** [reply]
Sorry I missed this. 
It's just a call to declareRestartableData in the constructor

**Reply by GiudGiud** [reply]
_grain_data is not a variable right? non linear variables are already restartable

**Comment by GiudGiud** [comment]
Hello

With the merging of [#22450](https://github.com/idaholab/moose/pull/22450) the situation should have improved here. 

Guillaume
